### PR TITLE
gh-94430: Allow parameters named `module` or `self` with custom C names in Argument Clinic

### LINKED ISSUE #94430: AC: Params named `module` with custom C names are rejected
**Feature or enhancement**

Currently, Argument Clinic will not allow parameters named `module`, even though they've got a custom C name. For example, this clinic input is rejected:

```c
/*[clinic input]
mod.func

    module as mod: object

[clinic start generated code]*/
```

AC complains with _"You can't have two parameters named 'module'!"_, even though there is no naming collision on the C level. This issue prevents AC conversion of some functions and methods in the code base, for example `warn_explicit()` in Python/_warnings.c. A workaround is to provide a custom name to the self converter, however I view that as a hack; it adds unneeded clinic code.

<details>
<summary>Hack/workaround</summary>

```c
/*[clinic input]
mod.func

    _module: self(type="PyObject *")
    module as mod: object

[clinic start generated code]*/
```

</details>

**Pitch**

The fix is easy (its _almost_ a one-liner), and it unlocks more AC conversions, without requiring unneeded hacks.

<details>
<summary>Patch</summary>

[diff.txt](https://github.com/python/cpython/files/9014411/diff.txt)

</details>

### PR DESCRIPTION
**Pull request by erlend-aasland**
* Issue: gh-94430

### CODE REVIEW

**Comment by erlend-aasland** [comment]
cc. @arhadthedev

**Comment by erlend-aasland** [comment]
Unfortunately, the fix is not as simple as this.

**Comment by erlend-aasland** [comment]
@JelleZijlstra, what do you think? Is this change worth it?

File: Tools/clinic/clinic.py [hunk 1]
```diff
@@ -4710,9 +4710,14 @@ def bad_node(self, node):
...
 
-        if parameter_name in self.function.parameters:
+        names = [k.name for k in self.function.parameters.values()]
+        if parameter_name in names[1:]:
             fail("You can't have two parameters named " + repr(parameter_name) + "!")
-        self.function.parameters[parameter_name] = p
+        elif names and parameter_name == names[0] and c_name is None:
+            fail("Params named 'module' or 'self' need custom C names")
+
+        key = f"{parameter_name}_{c_name}" if c_name else parameter_name
```
Comment (erlend-aasland) [inline Tools/clinic/clinic.py]: Possible improvement for future debugging of `self.function.parameters`; slightly improve key names:

```suggestion
        key = f"{parameter_name}_as_{c_name}" if c_name else parameter_name
```

File: Tools/clinic/clinic.py [hunk 2]
```diff
@@ -4710,9 +4710,14 @@ def bad_node(self, node):
 
         p = Parameter(parameter_name, kind, function=self.function, converter=converter, default=value, group=self.group)
 
-        if parameter_name in self.function.parameters:
+        names = [k.name for k in self.function.parameters.values()]
+        if parameter_name in names[1:]:
             fail("You can't have two parameters named " + repr(parameter_name) + "!")
-        self.function.parameters[parameter_name] = p
+        elif names and parameter_name == names[0] and c_name is None:
+            fail("Params named 'module' or 'self' need custom C names")
```
Comment (erlend-aasland) [inline Tools/clinic/clinic.py]: Maybe this is better:

```suggestion
            fail(f"Parameter '{parameter_name}' requires a custom C name")
```

Comment (erlend-aasland) [reply Tools/clinic/clinic.py]: Yes, it is better; I'll apply it.

**Comment by erlend-aasland** [comment]
IIRC, we backport AC changes. Adding 3.11 and 3.10 labels.

**Comment by erlend-aasland** [comment]
A pair of hearts and no thumbs down hints to some enthusiasm 🙂 Unless someone complains, I'll merge this next week. Reviews are still welcome.

**Review by kumaraditya303 — COMMENTED** [review]
LGTM, this would be useful in https://github.com/python/cpython/pull/92891

**Comment by erlend-aasland** [comment]
Thanks for the review, Kumar.

**Pull request merged**